### ISSUE TITLE
SQS receiveMessage not handling new lines correctly

### ISSUE THREAD

**Issue by pkmnfrk** [aws/aws-sdk-js-v3#1893]
**Describe the bug**
The SQS client seems to mishandle the first(?) new line in a message body, turning it from \n into a space. This causes the MD5 hash validation to fail and the receiveMessages call to throw.

```console
Error: Invalid MD5 checksum on messages: 12fcd5a6-338f-485a-9b01-949ad118a6f0
    at /home/mike/code/project/service/node_modules/@aws-sdk/middleware-sdk-sqs/dist/cjs/receive-message.js:21:19
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async /home/mike/code/project/service/node_modules/@aws-sdk/middleware-logger/dist/cjs/loggerMiddleware.js:7:22
```

**SDK version number**
~@aws-sdk/client-sqs@1.0.0-rc.10~
Reproduced in @aws-sdk/client-sqs@3.2.0

**Is the issue in the browser/Node.js/ReactNative?**
Node.js

**Details of the browser/Node.js/ReactNative version**
v14.8.0

**To Reproduce (observed behavior)**
Send a message to an SQS queue with a new line in it. Eg:
```json
{
    "hello": "world"
}
```

Then, receive that message in the code:

```js
const AWS = require("@aws-sdk/client-sqs");
const sqs = new AWS.SQS();
sqs
  .receiveMessage({
    QueueUrl: "https://sqs.us-east-1.amazonaws.com/...",
    WaitTimeSeconds: 20,
  })
  .then((data) => {
    console.log("Got", data.length, "messages");
  })
  .catch((e) => {
    console.error(e);
  });

```

**Expected behavior**
The call succeeds, and gives us the message.

**Comment by pkmnfrk** [comment]
I did a little more digging into this, and I found some surprising results.

![image](https://user-images.githubusercontent.com/1333816/104206724-dfb5d880-53fd-11eb-92c6-fac42af30499.png)

Top is the original message, as sent to SQS. You can see that the second byte is 0x0A (aka, \n). Additionally, it contains a trailing 0x0A.

Bottom is the message as received by SQS, obtained by adding a logger to the SQS client, and having it save to a file. The second byte is 0x20 (aka, space). It is also missing the trailing 0x0A. Calculating the hash of each file yields different results, of course, and the actual API response from SQS matches the hash of the top file.

```
mike@Mikescomp:~/code/project/service$ md5sum testbody2.json
d0b8560f261410878a68bbe070d81853  testbody2.json
mike@Mikescomp:~/code/project/service$ md5sum testbody.json
503d800db75e57a2e670034f30621032  testbody.json
```

![image](https://user-images.githubusercontent.com/1333816/104208203-8a2dfb80-53fe-11eb-8d56-2f4f7701c24d.png)

**Comment by axiac** [comment]
It trims the spaces present in `<Body>` on XML parsing.

The cause is in [`Aws_query.ts`](../blob/fa33f0b04b6c61e8305f07107f92da306d67242a/clients/client-sqs/protocols/Aws_query.ts#L3144):

```typescript
      const parsedObj = xmlParse(encoded, {
        attributeNamePrefix: "",
        ignoreAttributes: false,
        parseNodeValue: false,
        tagValueProcessor: (val, tagName) => decodeEscapedXML(val),
      });
```

It needs the option `trimValues: true`. The [default value](https://github.com/NaturalIntelligence/fast-xml-parser/blob/396ab96f8bd1f9953cfeff6c6b630396dd5f045a/src/xmlstr2xmlnode.js#L32) of `trimValues` is `false`.

**Comment by mmatheson** [comment]
Is there a workaround for this? In cases where the SQS queue is populated by an SNS topic we don't have control over the message structure, and there will be newlines in the message written by SNS ([example message here](https://docs.aws.amazon.com/sns/latest/dg/sns-sqs-as-subscriber.html)).

**Comment by matwerber1** [comment]
Seeing same issue. Looks like a regression? 

Works fine with SDK version 2.868.0, but having same issue as above when trying with new SDK v3.

```js
const AWS = require("aws-sdk"); // v 2.868.0
const sqs = new AWS.SQS({region: 'us-west-2'});

sqs.receiveMessage({
    QueueUrl: "https://sqs.us-west-2.amazonaws.com/.....",
    WaitTimeSeconds: 20,
}, (err,data) => {
    console.log(err);
    console.log(data);
});
```

Response:

```json
{
  ResponseMetadata: { RequestId: 'bab08679-6f3a-5168-a0ab-dea8a2fcc9d4' },
  Messages: [
    {
      MessageId: 'b6015f49-05c7-47e2-bbec-27b53d9bf31d',
      ReceiptHandle: 'xxxxx',
      MD5OfBody: '31d3a8830dfb4dd8009700b6a2c379d5',
      Body: '{\n    "hello": "world"\n}'
    }
  ]
}
```

**Comment by aduca98** [comment]
Any updates on this issue? I have been struggling for a day in a half with this issue and stumbled upon this thread and this is the exact issue I am having as well.

**Comment by redvivi** [comment]
Same here.

**Comment by blumamir** [comment]
Us too. After upgrading sdk from v2.780.0 to v3.8.0, the log is full of "Invalid MD5 checksum on messages" records, and the dead letter queue is stuffed with unhandled sqs messages.

This bug prevents us from using v3, as it effectively means that sqs client is not reliably able to receive messages.

**Comment by trivikr** [comment]
I'm not able to reproduce this issue while explicitly sending message with `\n` in them.

```console
$ node -v
v14.15.4
```

<details>
<summary>package.json</summary>

```json
"dependencies": {
  "@aws-sdk/client-sqs": "3.10.0",
  "md5": "2.3.0"
}
```

</details>

<details>
<summary>Code</summary>

```js
import { SQS } from "@aws-sdk/client-sqs";
import md5 from "md5";

const region = "us-west-2";
const client = new SQS({ region });

const QueueName = `test-${Date.now()}`;
const { QueueUrl } = await client.createQueue({ QueueName });

const MessageBody = `{\n  "hello": "world"\n}`;
console.log(MessageBody);
const { MD5OfMessageBody } = await client.sendMessage({
  MessageBody,
  QueueUrl,
});

console.log(
  `\nmd5 check for sent message: ${md5(MessageBody) === MD5OfMessageBody}`
);

const response = await client.receiveMessage({ QueueUrl });
console.log(
  `md5 check for received message: ${
    md5(MessageBody) === md5(response.Messages[0].Body)
  }`
);
console.log(
  `\nStrict Equality Comparison: ${MessageBody === response.Messages[0].Body}`
);

await client.deleteQueue({ QueueUrl });

```

</details>

<details>
<summary>Output</summary>

```console
$ node sqs.mjs
{
  "hello": "world"
}

md5 check for sent message: true
md5 check for received message: true

Strict Equality Comparison: true
```

</details>

Does this minimal repro code need any change to reproduce the issue?
@pkmnfrk @axiac @mmatheson @matwerber1 @aduca98 @blumamir

**Comment by trivikr** [comment]
The issue is reproducible if message begins with `\n` as follows:

```console
$ node -v
v14.15.4
```

<details>
<summary>package.json</summary>

```json
"dependencies": {
  "@aws-sdk/client-sqs": "3.10.0",
  "md5": "2.3.0"
}
```

</details>

<details>
<summary>Code</summary>

```js
import { SQS } from "@aws-sdk/client-sqs";
import md5 from "md5";

const region = "us-west-2";
const client = new SQS({ region });

const QueueName = `test-${Date.now()}`;
const { QueueUrl } = await client.createQueue({ QueueName });

const MessageBody = `\n{"hello": "world"}`;
console.log(MessageBody);
const { MD5OfMessageBody } = await client.sendMessage({
  MessageBody,
  QueueUrl,
});

console.log(
  `\nmd5 check for sent message: ${md5(MessageBody) === MD5OfMessageBody}`
);

const response = await client.receiveMessage({ QueueUrl });
console.log(
  `md5 check for received message: ${
    md5(MessageBody) === md5(response.Messages[0].Body)
  }`
);
console.log(
  `\nStrict Equality Comparison: ${MessageBody === response.Messages[0].Body}`
);

await client.deleteQueue({ QueueUrl });

```

</details>

<details>
<summary>Output</summary>

```console
$ node sqs.mjs

{"hello": "world"}

md5 check for sent message: true
/Users/trivikr/workspace/test/node_modules/@aws-sdk/middleware-sdk-sqs/dist/cjs/receive-message.js:21
            throw new Error("Invalid MD5 checksum on messages: " + messageIds.join(", "));
                  ^

Error: Invalid MD5 checksum on messages: d1ddc9da-4602-430b-9d15-82b6bcc4a7f0
    at /Users/trivikr/workspace/test/node_modules/@aws-sdk/middleware-sdk-sqs/dist/cjs/receive-message.js:21:19
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async /Users/trivikr/workspace/test/node_modules/@aws-sdk/middleware-logger/dist/cjs/loggerMiddleware.js:6:22
    at async file:///Users/trivikr/workspace/test/sqs.mjs:21:18
```

</details>

Verified that this issue doesn't reproduce with `aws-sdk@2.875.0`

**Comment by trivikr** [comment]
The line of code which throws error:
https://github.com/aws/aws-sdk-js-v3/blob/62ea7873559130d075a3a01082a1addda8aac1bf/packages/middleware-sdk-sqs/src/receive-message.ts#L43-L45

If I uncomment the lines in middleware-sdk-sqs and run code, the output is:
```

{"hello": "world"}

md5 check for sent message: true
md5 check for received message: false

Strict Equality Comparison: false
```

This happens as the first character `\n` is missing from received message:
* MessageBody: `\n{"hello": "world"}`
* response.Messages[0].Body: `{"hello": "world"}`

**Comment by trivikr** [comment]
Verified that the issue is with fast-xml-parser by logging data before (i.e. `encoded`) and after (i.e. `parsedObj?.ReceiveMessageResponse?.ReceiveMessageResult?.Message`) parsing on line https://github.com/aws/aws-sdk-js-v3/blob/fa33f0b04b6c61e8305f07107f92da306d67242a/clients/client-sqs/protocols/Aws_query.ts#L3144-L3149

Note that `\n` is present in encoded:
```
<?xml version="1.0"?><ReceiveMessageResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><ReceiveMessageResult><Message><MessageId>72078a1f-d069-4227-8ef7-3782a22a41dd</MessageId><ReceiptHandle>AQEBwo40ueEiRowuOwWTm4UTEzZk8w3U2JS19MnrlPJg5L4tlyZC1Zfa+gDuf4OnlvaGYA2kcl1i1QfnwkspJ077XUIuGNLPZYP+rZusK/4kEKANnP8XH9hHHXvPt4KKBu7uBFyZeTpRj6PzjpMPQhgNs2GTqFo/0prhchgOBcZAvwhtBKlPy3kNjC9bfpnOqorJC5A5CWdjt3zH1DjuG/RFVnTYkUZaebVWHf1yCSwj9paqOUPVjnzI3zWNS8LtEm6OmG+1wQ5Nqre8wZX2LHNzdMCK9gZmiqP3DTckykWwG/0OGxRmnMHT5IugHxU4LpUM92lDh1++4g+lhNGAnaLo3nY+MtnDXYFCRV5f4t0YEXIsxQrice4JmyUT8fFVrj7F1+P2r+IFuLPY/r4PPbrJ3A==</ReceiptHandle><MD5OfBody>a0a8c05f0dae22191973b61bb084a8b4</MD5OfBody><Body>
{&quot;hello&quot;: &quot;world&quot;}</Body></Message></ReceiveMessageResult><ResponseMetadata><RequestId>db286b83-7245-51b0-8424-335fdc5833ed</RequestId></ResponseMetadata></ReceiveMessageResponse>
```

However, it's removed in processed Message:
```
{
  MessageId: '72078a1f-d069-4227-8ef7-3782a22a41dd',
  ReceiptHandle: 'AQEBwo40ueEiRowuOwWTm4UTEzZk8w3U2JS19MnrlPJg5L4tlyZC1Zfa+gDuf4OnlvaGYA2kcl1i1QfnwkspJ077XUIuGNLPZYP+rZusK/4kEKANnP8XH9hHHXvPt4KKBu7uBFyZeTpRj6PzjpMPQhgNs2GTqFo/0prhchgOBcZAvwhtBKlPy3kNjC9bfpnOqorJC5A5CWdjt3zH1DjuG/RFVnTYkUZaebVWHf1yCSwj9paqOUPVjnzI3zWNS8LtEm6OmG+1wQ5Nqre8wZX2LHNzdMCK9gZmiqP3DTckykWwG/0OGxRmnMHT5IugHxU4LpUM92lDh1++4g+lhNGAnaLo3nY+MtnDXYFCRV5f4t0YEXIsxQrice4JmyUT8fFVrj7F1+P2r+IFuLPY/r4PPbrJ3A==',
  MD5OfBody: 'a0a8c05f0dae22191973b61bb084a8b4',
  Body: '{"hello": "world"}'
}
```

**Comment by trivikr** [comment]
This is an issue with upstream dependency fast-xml-parser, and the bug report is cut at https://github.com/NaturalIntelligence/fast-xml-parser/issues/331

**Comment by trivikr** [comment]
Verified that this issue is fixed in `v3.11.0`:

```console
$ node -v
v14.16.0
```

<details>
<summary>package.json</summary>

```json
"dependencies": {
  "@aws-sdk/client-sqs": "3.11.0",
  "md5": "2.3.0"
}
```

</details>

<details>
<summary>Code</summary>

```js
import { SQS } from "@aws-sdk/client-sqs";
import md5 from "md5";

const region = "us-west-2";
const client = new SQS({ region });

const QueueName = `test-${Date.now()}`;
const { QueueUrl } = await client.createQueue({ QueueName });

const MessageBody = `\n{"hello": "world"}`;
console.log(MessageBody);
const { MD5OfMessageBody } = await client.sendMessage({
  MessageBody,
  QueueUrl,
});

console.log(
  `\nmd5 check for sent message: ${md5(MessageBody) === MD5OfMessageBody}`
);

const response = await client.receiveMessage({ QueueUrl });
console.log(
  `md5 check for received message: ${
    md5(MessageBody) === md5(response.Messages[0].Body)
  }`
);
console.log(
  `\nStrict Equality Comparison: ${MessageBody === response.Messages[0].Body}`
);

console.log(response.Messages[0].Body);

await client.deleteQueue({ QueueUrl });

```

</details>

<details>
<summary>Output</summary>

```console
$ node sqs.mjs

{"hello": "world"}

md5 check for sent message: true
md5 check for received message: true

Strict Equality Comparison: true

{"hello": "world"}
```

</details>
